### PR TITLE
Cr 3692 azure context postgre sql server

### PR DESCRIPTION
**Pull request by lirshindalman**

### CODE REVIEW

File: cloudrail/knowledge/context/azure/azure_environment_context.py [hunk 1]
```diff
@@ -13,7 +14,8 @@
 from cloudrail.knowledge.context.azure.resources.databases.azure_cosmos_db_account import AzureCosmosDBAccount
 from cloudrail.knowledge.context.azure.resources.databases.azure_mssql_server_extended_auditing_policy import AzureSqlServerExtendedAuditingPolicy
 from cloudrail.knowledge.context.azure.resources.databases.azure_mysql_server import AzureMySqlServer
-from cloudrail.knowledge.context.azure.resources.databases.azure_postgresql_server import AzurePostgreSqlServer
+from cloudrail.knowledge.context.azure.resources.databases.azure_postgresql_server import AzurePostgreSqlServer, \
+    AzurePostgreSqlServerConfiguration
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/azure_environment_context.py:18]: I think that "AzurePostgreSqlServerConfiguration" should be imported from it's own file, currently it's being imported from the same file as "AzurePostgreSqlServer"

File: cloudrail/knowledge/context/azure/azure_relations_assigner.py [hunk 2]
```diff
@@ -97,6 +102,16 @@ def _assign_config_to_function_app(function_app: AzureFunctionApp, app_service_c
         )
         function_app.app_service_config = app_service_config
 
+    @staticmethod
+    def _assign_config_to_postgresql_server(postgresql_server: AzurePostgreSqlServer,
+                                            postgresql_server_configs: AliasesDict[AzurePostgreSqlServerConfiguration]):
+        postgresql_server_configs = ResourceInvalidator.get_by_logic(
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/azure_relations_assigner.py:108]: probably line 113 could be removed, and we can just put here:
"postgresql_server.postgresql_configuration = ResourceInvalidator.get_by_logic( ..."

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py [hunk 3]
```diff
@@ -1,30 +1,122 @@
...
+                 auto_grow_enabled: bool,
+                 backup_retention_days: Optional[int],
+                 geo_redundant_backup_enabled: Optional[bool],
+                 identity: Optional[PostgreSqlServerIdentity],
+                 infrastructure_encryption_enabled: bool,
+                 public_network_access_enabled: bool,
+                 ssl_enforcement_enabled: bool,
+                 ssl_minimal_tls_version_enforced: str,
+                 storage_mb: Optional[int],
+                 tags: Dict[str, str] = None):
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py:91]: remove "tags" it's not needed in Azure resources, we collect it automatically in the builders.
Did you make sure that all fields which are "Optional" are indeed Optional? Each one of them might be None in some cases?

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py [hunk 4]
```diff
@@ -1,30 +1,122 @@
...
         super().__init__(AzureResourceType.AZURERM_POSTGRESQL_SERVER)
-        self.server_name: str = server_name
+        self.name: str = server_name
         self.with_aliases(server_name)
+        if tags:
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py:95]: remove tags

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py [hunk 5]
```diff
@@ -40,5 +132,6 @@ def is_tagable(self) -> bool:
         return True
 
     def to_drift_detection_object(self) -> dict:
-        return {'tags': self.tags, 'server_name': self.server_name,
+        return {'tags': self.tags,
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py:135]: Should we add more fields here?
Are there anything else the user might change without re-creating the resource?

File: cloudrail/knowledge/context/azure/resources_builders/scanner/postgresql_server_builder.py [hunk 6]
```diff
@@ -8,5 +9,21 @@ def get_file_name(self) -> str:
...
+                                     administrator_login=properties.get('administratorLogin'),
+                                     auto_grow_enabled=properties.get('storageProfile').get('storageAutogrow') == 'Enabled',
+                                     backup_retention_days=properties.get('storageProfile').get('backupRetentionDays'),
+                                     geo_redundant_backup_enabled=properties.get('storageProfile').get('geoRedundantBackup') == 'Enabled',
+                                     identity=identity_type,
+                                     infrastructure_encryption_enabled=properties.get('infrastructureEncryption') == 'Enabled',
+                                     public_network_access_enabled=properties.get('publicNetworkAccess') == 'Enabled',
+                                     ssl_minimal_tls_version_enforced=properties.get('minimalTlsVersion', 'TLSEnforcementDisabled'),
+                                     storage_mb=properties.get('storageProfile').get('storageMB'),
+                                     tags=attributes.get('tags'))
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources_builders/scanner/postgresql_server_builder.py:29]: remove tags

File: cloudrail/knowledge/context/azure/resources_builders/terraform/postgresql_server_builder.py [hunk 7]
```diff
@@ -1,14 +1,29 @@
...
+                                     administrator_login=self._get_known_value(attributes, 'administrator_login'),
+                                     auto_grow_enabled=self._get_known_value(attributes, 'auto_grow_enabled', True),
+                                     backup_retention_days=self._get_known_value(attributes, 'backup_retention_days'),
+                                     geo_redundant_backup_enabled=self._get_known_value(attributes, 'geo_redundant_backup_enabled'),
+                                     identity=identity_type,
+                                     infrastructure_encryption_enabled=self._get_known_value(attributes, 'infrastructure_encryption_enabled', False),
+                                     public_network_access_enabled=self._get_known_value(attributes, 'public_network_access_enabled', True),
+                                     ssl_minimal_tls_version_enforced=self._get_known_value(attributes, 'ssl_minimal_tls_version_enforced', 'TLSEnforcementDisabled'),
+                                     storage_mb=self._get_known_value(attributes, 'storage_mb'),
+                                     tags=self._get_known_value(attributes, 'tags'))
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources_builders/terraform/postgresql_server_builder.py:26]: remove tags

File: cloudrail/knowledge/context/azure/terraform/azure_terraform_fields.yaml [hunk 8]
```diff
@@ -237,6 +237,32 @@ azurerm_postgresql_server:
...
+      - administrator_login
+      - administrator_login_password
+      - auto_grow_enabled
+      - backup_retention_days
+      - geo_redundant_backup_enabled
+      - infrastructure_encryption_enabled
+      - public_network_access_enabled
+      - ssl_minimal_tls_version_enforced
+      - storage_mb
+      - tags:
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/terraform/azure_terraform_fields.yaml:251]: remove tags

File: tests/knowledge/context/azure/test_postgresql_server.py [hunk 9]
```diff
@@ -20,3 +21,33 @@ def test_postgresql_enforcing_ssl_not_enabled(self, ctx: AzureEnvironmentContext
         server = ctx.postgresql_servers.get('cr2467-postgresql-server')
         self.assertIsNotNone(server)
         self.assertFalse(server.ssl_enforcement_enabled)
+
+    @context(module_path="basic", test_options=TestOptions(run_drift_detection=False, run_cloudmapper=False))
```
Comment (Imanuel-Mizrahi) [inline tests/knowledge/context/azure/test_postgresql_server.py:25]: remove the test options once you can

**Review by Imanuel-Mizrahi — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server_configuration.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+            value: Specifies the value of the PostgreSQL Configuration.
+    """
+
+    def __init__(self,
+                 name: str,
+                 server_name: str,
+                 value: str):
+        super().__init__(AzureResourceType.AZURERM_POSTGRESQL_SERVER_CONFIGURATION)
+        self.name: str = name
+        self.with_aliases(f'{name}_{server_name}')
```
Comment (Tomerva) [inline cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server_configuration.py:20]: i dont think this is necessary as you dont use it, and (probably) your only usage for the collection of psql configs will be to assign them to the proper psql server, which doesnt require fast access to the configs (but does to the servers, which you already have)

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py [hunk 11]
```diff
@@ -1,30 +1,119 @@
...
+            disabled_alerts: List of alerts which should be disabled.
+            email_account_admins: If account administrators should be emailed when an alert is triggered.
+            email_addresses: A list of email addresses which alerts should be sent to.
+            retention_days: Specifies the number of days to keep in the Threat Detection audit logs.
+            storage_account_access_key: Specifies the identifier key of the Threat Detection audit storage account.
+            storage_endpoint: Specifies the blob storage endpoint. This blob storage will hold all Threat Detection audit logs.
+    """
+    enabled: bool
+    disabled_alerts: Optional[List[PostgreSqlServerThreatDetectionPolicyAlerts]]
+    email_account_admins: Optional[bool]
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py:52]: please note all the optional fields, in most cases, bool is not an optional value.
In addition, I cannot see this class being used anywhere...
Perhaps we can remove it?

**Review by Imanuel-Mizrahi — APPROVED** [review]

File: cloudrail/knowledge/context/azure/azure_relations_assigner.py [hunk 12]
```diff
@@ -97,6 +102,15 @@ def _assign_config_to_function_app(function_app: AzureFunctionApp, app_service_c
         )
         function_app.app_service_config = app_service_config
 
+    @staticmethod
+    def _assign_config_to_postgresql_server(postgresql_server: AzurePostgreSqlServer,
+                                            postgresql_server_configs: AliasesDict[AzurePostgreSqlServerConfiguration]):
+        postgresql_server.postgresql_configuration = ResourceInvalidator.get_by_logic(
```
Comment (Tomerva) [inline cloudrail/knowledge/context/azure/azure_relations_assigner.py:108]: is it possible that psql doesnt have config? if not, it should be invalidated

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py [hunk 13]
```diff
@@ -40,5 +129,5 @@ def is_tagable(self) -> bool:
         return True
 
     def to_drift_detection_object(self) -> dict:
-        return {'tags': self.tags, 'server_name': self.server_name,
```
Comment (Tomerva) [inline cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py:43]: why removed tags?

**Review by Tomerva — CHANGES_REQUESTED** [review]

File: cloudrail/knowledge/context/azure/azure_relations_assigner.py (hunk 12 above)
Comment (lirshindalman) [reply cloudrail/knowledge/context/azure/azure_relations_assigner.py:108]: its possible that psql doesnt have config

File: cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py (hunk 13 above)
Comment (lirshindalman) [reply cloudrail/knowledge/context/azure/resources/databases/azure_postgresql_server.py:43]: "tags" not needed in Azure resources, we collect it automatically in the builders.

**Review by Tomerva — APPROVED** [review]

**Pull request merged**